File: scripts/lib/extractors/python.py
"""Python via tree-sitter-python. Namespace derived from file path."""
from __future__ import annotations
from pathlib import Path
import tree_sitter_python as _grammar
from tree_sitter import Language, Parser, Query

from .base import Declaration, ImportSpec, ParseResult
from ._common import text_of, has_error_in, walk, run_query, loc_of, signature_of, count_methods

name = "python"
extensions = (".py",)
grammar_package = "tree-sitter-python"

_LANG = Language(_grammar.language())
_PARSER = Parser(_LANG)

_Q_IMPORT = Query(_LANG, """
(import_statement (dotted_name) @import)
(import_from_statement module_name: (dotted_name) @import)
(import_from_statement module_name: (relative_import) @import)
""")
# Top-level declarations only. `module` is the root grammar node.
_Q_DECL = Query(_LANG, """
(module (class_definition name: (identifier) @name) @decl)
(module (function_definition name: (identifier) @name) @decl)
(module (decorated_definition (class_definition name: (identifier) @name)) @decl)
(module (decorated_definition (function_definition name: (identifier) @name)) @decl)
""")


def _path_to_namespace(rel: str) -> str:
    """src/api/routes/order.py → src.api.routes.order  (the file itself is the namespace tail)"""
    p = Path(rel)
    parts = list(p.parts[:-1]) + [p.stem]
    # Drop trailing __init__ for packages
    if parts and parts[-1] == "__init__":
        parts = parts[:-1]
    return ".".join(parts) if parts else ""


def _inner(decl_node):
    """Unwrap a decorated_definition to the class/function it decorates."""
    if decl_node.type == "decorated_definition":
        for c in decl_node.children:
            if c.type in ("class_definition", "function_definition"):
                return c
    return decl_node


def _kind(decl_node):
    return "class" if _inner(decl_node).type == "class_definition" else "function"


def _supertypes(decl_node, src):
    """class A(B, C, metaclass=M) → ['B', 'C']  (metaclass kwargs ignored)"""
    inner = decl_node
    if decl_node.type == "decorated_definition":
        for c in decl_node.children:
            if c.type == "class_definition":
                inner = c; break
        else:
            return [], "high"
    if inner.type != "class_definition":
        return [], "high"
    out, conf = [], "high"
    for c in inner.children:
        if c.type == "argument_list":
            if has_error_in(c):
                conf = "low"; continue
            for arg in c.children:
                if arg.type == "identifier":
                    out.append(text_of(arg, src))
                elif arg.type == "attribute":  # foo.Bar
                    out.append(text_of(arg, src))
    return out, conf


def parse(path: Path, src: bytes, project_root: Path) -> ParseResult:
    rel = str(path.relative_to(project_root))
    namespace = _path_to_namespace(rel)
    root = _PARSER.parse(src).root_node

    imports = []
    for caps in run_query(_Q_IMPORT, root):
        for n in caps.get("import", []):
            raw = text_of(n, src)
            imports.append(ImportSpec(raw=raw, qualified=raw, alias=raw.rsplit(".", 1)[-1]))

    decls, skipped = [], []
    for caps in run_query(_Q_DECL, root):
        decl = caps["decl"][0]
        name_node = caps["name"][0]
        if has_error_in(name_node):
            skipped.append({"path": rel, "line": decl.start_point[0]+1, "reason": "name_errored"})
            continue
        supers, conf = _supertypes(decl, src)
        inner = _inner(decl)
        kind = _kind(decl)
        decls.append(Declaration(
            name=text_of(name_node, src),
            namespace=namespace or None,
            kind=kind,
            path=rel,
            line=decl.start_point[0] + 1,
            supertypes=supers,
            refs=[i.qualified for i in imports if i.qualified],
            confidence=conf,
            loc=loc_of(decl),
            signature=signature_of(inner, src) if kind == "function" else "",
            method_count=count_methods(inner, ("block",), ("function_definition", "decorated_definition")) if kind == "class" else 0,
        ))
    return ParseResult(declarations=decls, imports=imports, skipped=skipped)

File: scripts/lib/extractors/python.mjs
// scripts/lib/extractors/python.mjs — Python via tree-sitter-python.
// Faithful port of scripts/lib/extractors/python.py. Namespace from file path.
import { init, loadLanguage, makeQuery, Parser } from '../ts.mjs';
import { Declaration, ImportSpec, ParseResult } from './base.mjs';
import { textOf, hasErrorIn, walk, runQuery, locOf, signatureOf, countMethods } from './_common.mjs';

export const name = 'python';
export const extensions = ['.py'];
export const grammar = 'python'; // manifest key

let _lang, _qImport, _qDecl;
async function ensure() {
  if (_lang) return;
  await init();
  _lang = await loadLanguage('python');
  _qImport = makeQuery(_lang, `
(import_statement (dotted_name) @import)
(import_from_statement module_name: (dotted_name) @import)
(import_from_statement module_name: (relative_import) @import)
`);
  _qDecl = makeQuery(_lang, `
(module (class_definition name: (identifier) @name) @decl)
(module (function_definition name: (identifier) @name) @decl)
(module (decorated_definition (class_definition name: (identifier) @name)) @decl)
(module (decorated_definition (function_definition name: (identifier) @name)) @decl)
`);
}

// src/api/routes/order.py → src.api.routes.order  (file itself is the namespace tail)
function pathToNamespace(rel) {
  const segs = rel.split('/');
  const file = segs.pop();
  const stem = file.replace(/\.[^.]+$/, ''); // drop final extension
  const parts = [...segs, stem];
  if (parts.length && parts[parts.length - 1] === '__init__') parts.pop();
  return parts.length ? parts.join('.') : '';
}

// Single leading underscore (and not a __dunder__) = module-private by Python
// convention. Lets core.mjs downweight internal helpers out of `core`.
function pyVisibility(name) {
  return /^_/.test(name) && !/^__.*__$/.test(name) ? 'private' : 'public';
}

function inner(declNode) {
  if (declNode.type === 'decorated_definition') {
    for (const c of declNode.children) {
      if (c.type === 'class_definition' || c.type === 'function_definition') return c;
    }
  }
  return declNode;
}

function kindOf(declNode) {
  return inner(declNode).type === 'class_definition' ? 'class' : 'function';
}

function supertypesOf(declNode, src) {
  let inr = declNode;
  if (declNode.type === 'decorated_definition') {
    let found = null;
    for (const c of declNode.children) { if (c.type === 'class_definition') { found = c; break; } }
    if (found == null) return [[], 'high'];
    inr = found;
  }
  if (inr.type !== 'class_definition') return [[], 'high'];
  const out = [];
  let conf = 'high';
  for (const c of inr.children) {
    if (c.type === 'argument_list') {
      if (hasErrorIn(c)) { conf = 'low'; continue; }
      for (const arg of c.children) {
        if (arg.type === 'identifier' || arg.type === 'attribute') out.push(textOf(arg, src));
      }
    }
  }
  return [out, conf];
}

function bodyRefs(declNode, src) {
  const names = new Set();
  for (const n of walk(declNode)) {
    if (n.type !== 'call') continue;
    const fn = n.childForFieldName('function');
    if (fn == null) continue;
    if (fn.type === 'identifier') {
      names.add(textOf(fn, src));
    } else if (fn.type === 'attribute') {
      const attr = fn.childForFieldName('attribute');
      if (attr != null && attr.type === 'identifier') names.add(textOf(attr, src));
    }
  }
  return [...names].sort();
}

export async function parse(relPath, src, _projectRoot) {
  await ensure();
  const namespace = pathToNamespace(relPath);
  const parser = new Parser();
  parser.setLanguage(_lang);
  const root = parser.parse(src).rootNode;

  const imports = [];
  for (const caps of runQuery(_qImport, root)) {
    for (const n of caps.import ?? []) {
      const raw = textOf(n, src);
      imports.push(ImportSpec(raw, raw, raw.split('.').pop()));
    }
  }

  const decls = [];
  const skipped = [];
  for (const caps of runQuery(_qDecl, root)) {
    const decl = caps.decl[0];
    const nameNode = caps.name[0];
    if (hasErrorIn(nameNode)) {
      skipped.push({ path: relPath, line: decl.startPosition.row + 1, reason: 'name_errored' });
      continue;
    }
    const [supers, conf] = supertypesOf(decl, src);
    const inr = inner(decl);
    const kind = kindOf(decl);
    const dname = textOf(nameNode, src);
    decls.push(Declaration({
      name: dname,
      namespace: namespace || null,
      kind,
      visibility: pyVisibility(dname),
      path: relPath,
      line: decl.startPosition.row + 1,
      supertypes: supers,
      refs: bodyRefs(inr, src),
      confidence: conf,
      loc: locOf(decl),
      signature: kind === 'function' ? signatureOf(inr, src) : '',
      method_count: kind === 'class'
        ? countMethods(inr, ['block'], ['function_definition', 'decorated_definition'])
        : 0,
    }));
  }
  return ParseResult(decls, imports, skipped);
}
